File: pset6/sentimental/credit/credit.py
# filename: credit.py
""" Get user input (credit card number) and validate if the card is legit.
    Also print which credit card company the card belongs to.
"""
# get user input and validate given input
while True:
    try:
        cc_number = int(input("Number: "))
    except ValueError:
        continue
    if cc_number >= 0:
        break

# change cc_number to string for indexing
cc_number = str(cc_number)
num_length = len(cc_number)

# get sum of every other number multiplied by 2, starting from the back
sum_1 = 0
for i in range(2, num_length + 1, 2):
    digit = int(cc_number[-i]) * 2
    # check if multiplied number is more than a digit long
    # if so, add each digit to sum_1, if not, add digit to sum_1
    if len(str(digit)) > 1:
        for j in range(len(str(digit))):
            sum_1 += int(str(digit)[j])
    else:
        sum_1 += int(digit)

# get sum of every other number that wasn't multiplied
sum_2 = 0
for i in range(1, num_length + 1, 2):
    sum_2 += int(cc_number[-i])

# check if card number is valid by using % or just by checking if last digit == 0
if (sum_2 + sum_1) % 10 == 0:
    amex = (34, 37)
    master = (51, 52, 53, 54, 55)

    # find which credit company the card belongs to
    if int(cc_number[:2]) in amex and num_length == 15:
        print("AMEX")
    elif int(cc_number[:2]) in master and num_length == 16:
        print("MASTERCARD")
    elif int(cc_number[0]) == 4 and num_length == 13 or num_length == 16:
        print("VISA")
    else:
        print("INVALID")
else:
    print('INVALID')

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long long cc_number;
    int sum = 0;
    int number_length = 0;
    int first_digit = 0;
    int second_digit = 0;
    do
    {
        cc_number = get_long_long("Number: ");
    }
    while (cc_number == 0 || cc_number < 0);
    for (long long x = 100; x < cc_number * 10; x *= 100)
    {
        number_length++;
        int digit = (cc_number % x) / (x/10);
        second_digit = digit;
        if (digit * 2 > 9)
        {
            sum += ((digit * 2) % 10) + (((digit * 2) % 100) / 10);
        }
        else
        {
            sum += digit * 2;
        }

    }
    for (long long x = 10; x < cc_number * 10; x *= 100)
    {
       number_length++;
       sum += (cc_number % x) / (x/10);
       first_digit = (cc_number % x) / (x/10);
    }
    if (sum % 10 == 0)
    {
        if (number_length % 2 == 0)
        {
            if (second_digit == 5 && first_digit >= 1 && first_digit <= 5)
            {
                printf("MASTERCARD\n");
            }
            else if (second_digit == 4 && number_length == 16)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            if (first_digit == 3 && (second_digit == 4 || second_digit == 7) && number_length == 15)
            {
                printf("AMEX\n");
            }
            else if (first_digit == 4 && number_length == 13)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
}